mtcpclient/calculate.go

```diff
@@ -16,48 +16,42 @@ type metricsCollectionStats struct {
 	numberOfConnections int
 }
 
-func (gc GroupOfConnections) calculateMetricsReport(status tcpclient.ConnectionStatus) (mr metricsCollectionStats) {
-	// TODO: There's something i don't like... initiatlizing values in the first loop, and the standard deviation
-	//  requiring an extra pass considering all items... i'd move initialization out of the loop, and maybe iterate
-	//  over a filtered list rather than several loops over the original one, and maybe use specific generic functions...
-	//  requires further thinking in any case
-	// TODO: (2) rather than calculate a metrics report with a filter on the connection status, lets just promote splitting
-	// group of connections and running these functions over the subsets...
+func newMetricsCollectionStats() *metricsCollectionStats {
+	mr := new(metricsCollectionStats)
+	mr.avg = 0
+	mr.min = time.Duration(time.Duration(tcpclient.DefaultDialTimeoutInMs)*time.Millisecond + 1)
+	mr.max = 0
 	mr.total = 0
+	mr.stdDev = 0
 	mr.numberOfConnections = 0
-	for _, item := range gc {
-		if item.GetConnectionStatus() == status {
-			mr.numberOfConnections++
-			if mr.total == 0 {
-				mr.total = item.GetTCPProcessingDuration()
-				mr.min = item.GetTCPProcessingDuration()
-				mr.max = item.GetTCPProcessingDuration()
-			} else {
-				mr.min = time.Duration(math.Min(float64(mr.min), float64(item.GetTCPProcessingDuration())))
-				mr.max = time.Duration(math.Max(float64(mr.max), float64(item.GetTCPProcessingDuration())))
-				mr.total += item.GetTCPProcessingDuration()
-			}
+	return mr
+}
+
+func (gc GroupOfConnections) calculateMetricsReport() (mr *metricsCollectionStats) {
+	mr = newMetricsCollectionStats()
+	if mr.numberOfConnections = len(gc.connections); mr.numberOfConnections > 0 {
+		for _, item := range gc.connections {
+			mr.min = time.Duration(math.Min(float64(mr.min), float64(item.GetTCPProcessingDuration())))
+			mr.max = time.Duration(math.Max(float64(mr.max), float64(item.GetTCPProcessingDuration())))
+			mr.total += item.GetTCPProcessingDuration()
 		}
+		mr.avg = mr.total / time.Duration(mr.numberOfConnections)
```

Comment:
I see something "inconsistent", but maybe there's a reason. In order to use the math functions Min/Max, you convert durations to float64, and after getting the result, to duration again. But here, you convert to duration a value and then, you divide, storing the result. Is just the behaviour of these mathematical functions different, and this last one preserve and return a duration?

Reply:
To get a time.Duration through a division, both types need to be consistent, if not: `(mismatched types time.Duration and int)`